packages/controllers/src/permissions/Permission.ts

```diff
@@ -0,0 +1,476 @@
+ *
+ * @template Key - The target key string to apply the constraint to.
+ */
+type ValidateTargetKey<Key extends string> = Key extends `${string}_*`
+  ? Key
+  : Key extends `${string}_`
+  ? never
+  : Key extends `${string}*`
+  ? never
+  : Key;
+
+/**
+ * A type extended by any function.
+ */
+type GenericFunction = (...args: any[]) => any;
+
+export type PermissionSpecificationConstraint = {
+  /**
+   * The target resource of the permission. In other words, at the time of
+   * writing, the RPC method name.
+   */
+  targetKey: string;
+
+  /**
+   * An array of the caveat types that may be added to instances of this
+   * permission.
+   */
+  allowedCaveats?: Readonly<NonEmptyArray<string>>;
+
+  /**
+   * The implementation of the restricted method that the permission
+   * corresponds to.
+   */
+  methodImplementation: GenericFunction;
+
+  /**
+   * The factory function used to get permission objects. Permissions returned
+   * by this function are presumed to valid, and they will not be passed to the
+   * validator function associated with this specification (if any). In other
+   * words, the factory function should validate the permissions it creates.
+   *
+   * If no factory is specified, the {@link Permission} constructor will be
+   * used, and the validator function (if specified) will be called on newly
+   * constructed permissions.
+   */
+  factory?: PermissionFactory<any, Record<string, unknown>>;
+
+  /**
+   * The validator function used to validate permissions of the associated type
+   * whenever they are mutated. The only way a permission can be legally mutated
+   * is when its caveats are modified by the permission controller.
+   *
+   * The validator should throw an appropriate JSON-RPC error if validation fails.
+   */
+  validator?: PermissionValidator<GenericPermission>;
+};
+
+/**
+ * Constraint for {@link PermissionSpecificationConstraint} objects that
+ * evaluates to `never` if the specification contains any invalid fields.
+ *
+ * @template Specification - The permission specification to validate.
+ */
+export type ValidatePermissionSpecification<Specification> =
+  Specification extends PermissionSpecificationConstraint
+    ? Specification['targetKey'] extends ValidateTargetKey<
+        Specification['targetKey']
+      >
+      ? Specification['methodImplementation'] extends ValidateRestrictedMethod<
+          Specification['methodImplementation']
+        >
+        ? Specification
+        : never
+      : never
+    : never;
+
+/**
+ * The specifications for all permissions and restricted methods supported by
+ * a particular {@link PermissionController}.
+ *
+ * @template Specifications - A union of all {@link PermissionSpecificationConstraint}
+ * types.
+ */
+export type PermissionSpecificationMap<
+  Specification extends PermissionSpecificationConstraint,
+> = {
+  [TargetKey in Specification['targetKey']]: Specification extends PermissionSpecificationConstraint & {
```

Comment:
Nit: We don't need to extend `PermissionSpecificationConstraint` twice.

I tried removing it here and got type errors because it's doubly-extended below in `ExtractPermissionSpecification` as well. Removing it from both results in no type errors.